core/templates/pages/skill-editor-page/skill-editor-page.component.spec.ts

```diff
@@ -71,12 +74,18 @@ describe('Skill editor page', function() {
   });
 
   it('should call confirm before leaving', function() {
+    spyOn(SkillEditorStateService, 'loadSkill').and.stub();
+    spyOn(UrlService, 'getSkillIdFromUrl').and.returnValue('skill_1');
     spyOn(UndoRedoService, 'getChangeCount').and.returnValue(10);
-    spyOn(window, 'addEventListener');
-    ctrl.setUpBeforeUnload();
-    ctrl.confirmBeforeLeaving({returnValue: ''});
-    expect(window.addEventListener).toHaveBeenCalledWith(
-      'beforeunload', ctrl.confirmBeforeLeaving);
+    spyOn(PreventPageUnloadEventService, 'addListener').and
+      .callFake((callback) => {
+        callback();
+      });
+
+    ctrl.$onInit();
+
+    expect(PreventPageUnloadEventService.addListener)
+      .toHaveBeenCalledWith(jasmine.any(Function));
```

Comment:
> Why use any here? Is it possible to check that `UndoRedoService.getChangeCount` was passed?

Since `UndoRedoService.getChangeCount` was passed in an anonymous function I don't think it can be tested.

Reply:
What if you passed the `UndoRedoService.getChangeCount` instead of `() => UndoRedoService.getChangeCount()`. Could it be tested then?

Reply:
Passing `UndoRedoService.getChangeCount` doesn't work, since the alert modal does not display even if changes are present.

Reply:
Sorry, I don't understand. Can you please elaborate on "since the alert modal does not display even if changes are present"? Also, what exactly do you mean by "Passing UndoRedoService.getChangeCount doesn't work", do you mean that the test fails or that the functionality breaks?

Reply:
The functionality breaks on passing `UndoRedoService.getChangeCount`  instead of `() => UndoRedoService.getChangeCount()` here. https://github.com/oppia/oppia/blob/a77d93e6e842cd0155aa50d62a57953881eca44e/core/templates/pages/skill-editor-page/skill-editor-page.component.ts#L96-L97
The save-prompt Modal is **NOT** displayed when the page is reloaded despite changes being present.
I'm unable to find out why this is occurring since I can test the `_preventPageUnloadEventHandler` function that is given below only if the function is triggered when I reload the page and when I reload the `_preventPageUnloadEventHandler` is not triggered.
https://github.com/oppia/oppia/blob/a77d93e6e842cd0155aa50d62a57953881eca44e/core/templates/services/prevent-page-unload-event.service.ts#L55-L68

Reply:
Hm, I see. I suspect that perhaps `UndoRedoService.getChangeCount` may not be called with the correct scope? Have you investigated this?

Reply:
I tried the following for testing the function. All of them seem to have failed.

1. ```
    expect(PreventPageUnloadEventService.addListener)
      .toHaveBeenCalledWith(
        UndoRedoService.getChangeCount());
2. ```
    expect(PreventPageUnloadEventService.addListener)
      .toHaveBeenCalledWith(
        UndoRedoService.getChangeCount);
3. ```
    expect(PreventPageUnloadEventService.addListener)
      .toHaveBeenCalledWith(
        UndoRedoService.getChangeCount.bind(UndoRedoService));

Reply:
Oh, I see. Thank you so much for checking! It seems like Jasmine is not able to compare complex objects like Functions directly using just `toHaveBeenCalledWith()`. In that case, this can be left as is (i.e. use `jasmine.any(Function)`).

However, I noticed that the test description (`should call confirm before leaving`) might be incorrect. The test currently just checks that `addListener` is called after the initialisation method and not that the confirm modal actually shows, right? If so, can you please correct that. Thanks!

Reply:
DONE